File: XMLCheck/Handler/Xml8Parser.cs
using System.Globalization;
using XMLCheck.Models;

namespace XMLCheck.Handler
{
    public class Xml8Parser : BaseXmlParser, IXmlParser
    {
        public string LoaiHoSo => "XML8";

        public object Parse(string xmlContent)
        {
            var model = new Xml8Model();

            // 1. Load XML (Base xử lý BOM + Parse)
            if (!Load(xmlContent))
                return model;

            // 2. Check root
            if (!IsRoot("CHI_TIEU_DU_LIEU_TOM_TAT_HO_SO_BENH_AN"))
                return model;

            // ===== Helper đọc trực tiếp từ Root =====
            string? S8(string k) =>
                Root!
                    .Descendants()
                    .FirstOrDefault(x => x.Name.LocalName == k)
                    ?.Value
                    ?.Trim();

            int? I8(string k) =>
                int.TryParse(S8(k), out var v) ? v : null;

            DateTime? DT8(string k)
            {
                var raw = S8(k);
                if (string.IsNullOrWhiteSpace(raw)) return null;

                string[] fmts = { "yyyyMMdd", "yyyyMMddHHmm", "yyyyMMddHHmmss" };
                return DateTime.TryParseExact(
                    raw,
                    fmts,
                    CultureInfo.InvariantCulture,
                    DateTimeStyles.None,
                    out var d
                ) ? d : null;
            }

            // ===== Mapping =====
            model.MA_LK = S8("MA_LK");
            model.MA_LOAI_KCB = S8("MA_LOAI_KCB");

            model.HO_TEN_CHA = S8("HO_TEN_CHA");
            model.HO_TEN_ME = S8("HO_TEN_ME");
            model.NGUOI_GIAM_HO = S8("NGUOI_GIAM_HO");
            model.DON_VI = S8("DON_VI");

            model.NGAY_VAO = DT8("NGAY_VAO");
            model.NGAY_RA = DT8("NGAY_RA");

            model.CHAN_DOAN_VAO = S8("CHAN_DOAN_VAO");
            model.CHAN_DOAN_RV = S8("CHAN_DOAN_RV");

            model.QT_BENHLY = S8("QT_BENHLY");
            model.TOMTAT_KQ = S8("TOMTAT_KQ");
            model.PP_DIEUTRI = S8("PP_DIEUTRI");

            model.NGAY_SINHCON = DT8("NGAY_SINHCON");
            model.NGAY_CONCHET = DT8("NGAY_CONCHET");
            model.SO_CONCHET = I8("SO_CONCHET");

            model.KET_QUA_DTRI = S8("KET_QUA_DTRI");
            model.GHI_CHU = S8("GHI_CHU");

            model.MA_TTDV = S8("MA_TTDV");
            model.NGAY_CT = DT8("NGAY_CT");

            model.MA_THE_TAM = S8("MA_THE_TAM");
            model.DU_PHONG = S8("DU_PHONG");

            return model;
        }
    }
}

File: XMLValidator_v2/src/parsers/parsers/xml8.parser.ts
import { BaseXmlParser } from '../base-xml.parser';
import { Xml8Model } from '../../interfaces/xml-models.interface';

export class Xml8Parser extends BaseXmlParser {
  public get loaiHoSo(): string {
    return 'XML8';
  }

  public parse(xmlContent: string): Xml8Model {
    const model: Xml8Model = {};

    if (!this.load(xmlContent)) return model;
    if (!this.isRoot('CHI_TIEU_DU_LIEU_TOM_TAT_HO_SO_BENH_AN')) return model;

    model.MA_LK = this.S('MA_LK') || undefined;
    model.MA_LOAI_KCB = this.S('MA_LOAI_KCB') || undefined;
    model.HO_TEN_CHA = this.S('HO_TEN_CHA') || undefined;
    model.HO_TEN_ME = this.S('HO_TEN_ME') || undefined;
    model.NGUOI_GIAM_HO = this.S('NGUOI_GIAM_HO') || undefined;
    model.DON_VI = this.S('DON_VI') || undefined;
    model.NGAY_VAO = this.DT('NGAY_VAO') ?? undefined;
    model.NGAY_RA = this.DT('NGAY_RA') ?? undefined;
    model.CHAN_DOAN_VAO = this.S('CHAN_DOAN_VAO') || undefined;
    model.CHAN_DOAN_RV = this.S('CHAN_DOAN_RV') || undefined;
    model.QT_BENHLY = this.S('QT_BENHLY') || undefined;
    model.TOMTAT_KQ = this.S('TOMTAT_KQ') || undefined;
    model.PP_DIEUTRI = this.S('PP_DIEUTRI') || undefined;
    model.NGAY_SINHCON = this.DT('NGAY_SINHCON') ?? undefined;
    model.NGAY_CONCHET = this.DT('NGAY_CONCHET') ?? undefined;
    model.SO_CONCHET = this.I('SO_CONCHET') ?? undefined;
    model.KET_QUA_DTRI = this.S('KET_QUA_DTRI') || undefined;
    model.GHI_CHU = this.S('GHI_CHU') || undefined;
    model.MA_TTDV = this.S('MA_TTDV') || undefined;
    model.NGAY_CT = this.DT('NGAY_CT') ?? undefined;
    model.MA_THE_TAM = this.S('MA_THE_TAM') || undefined;
    model.DU_PHONG = this.S('DU_PHONG') || undefined;

    return model;
  }
}
